src/document_basic.c

```diff
@@ -1,94 +1,181 @@
   f->indexAs = typemask;
-  f->name = strdup(fieldname);  // free_detached called on this later on..
+  f->name = fieldname;
   f->text = fieldval;
 }
 
-void Document_PrepareForAdd(Document *doc, RedisModuleString *docKey, double score,
-                            const AddDocumentOptions *opts, RedisModuleCtx *ctx) {
-  size_t payloadSize = 0;
-  const char *payloadStr = NULL;
-  if (opts->payload) {
-    payloadStr = RedisModule_StringPtrLen(opts->payload, &payloadSize);
+void Document_SetPayload(Document *d, const void *p, size_t n) {
+  d->payload = p;
+  d->payloadSize = n;
+  if (d->flags & DOCUMENT_F_OWNSTRINGS) {
+    d->payload = malloc(n);
+    memcpy((void *)d->payload, p, n);
   }
+}
 
-  Document_Init(doc, docKey, score, opts->numFieldElems / 2,
-                opts->language ? opts->language : DEFAULT_LANGUAGE, payloadStr, payloadSize);
-  for (size_t ii = 0, n = 0; ii < opts->numFieldElems; ii += 2, n++) {
-    // printf("indexing '%s' => '%s'\n", RedisModule_StringPtrLen(opts->fieldsArray[ii], NULL),
-    //        RedisModule_StringPtrLen(opts->fieldsArray[ii + 1], NULL));
-    doc->fields[n].name = RedisModule_StringPtrLen(opts->fieldsArray[ii], NULL);
-    doc->fields[n].text = RedisModule_CreateStringFromString(ctx, opts->fieldsArray[ii + 1]);
+void Document_Move(Document *dst, Document *src) {
+  if (dst == src) {
+    return;
   }
-
-  Document_Detach(doc, ctx);
+  *dst = *src;
+  src->flags |= DOCUMENT_F_DEAD;
 }
 
-void Document_DetachFields(Document *doc, RedisModuleCtx *ctx) {
-  for (size_t ii = 0; ii < doc->numFields; ++ii) {
-    DocumentField *f = doc->fields + ii;
+void Document_MakeStringsOwner(Document *d) {
+  if (d->flags & DOCUMENT_F_OWNSTRINGS) {
+    // Already the owner
+    return;
+  }
+  d->docKey = RedisModule_CreateStringFromString(RSDummyContext, d->docKey);
+  for (size_t ii = 0; ii < d->numFields; ++ii) {
+    DocumentField *f = d->fields + ii;
+    f->name = strdup(f->name);
     if (f->text) {
-      RedisModule_RetainString(ctx, f->text);
+      f->text = RedisModule_CreateStringFromString(RSDummyContext, f->text);
     }
-    f->name = strdup(f->name);
   }
+  if (d->payload) {
+    void *tmp = malloc(d->payloadSize);
+    memcpy(tmp, d->payload, d->payloadSize);
+    d->payload = tmp;
+  }
+  d->flags |= DOCUMENT_F_OWNSTRINGS;
 }
 
-void Document_ClearDetachedFields(Document *doc, RedisModuleCtx *anyCtx) {
-  for (size_t ii = 0; ii < doc->numFields; ++ii) {
-    if (doc->fields[ii].text) {
-      RedisModule_FreeString(anyCtx, doc->fields[ii].text);
+int Document_LoadSchemaFields(Document *doc, RedisSearchCtx *sctx) {
+  RedisModuleKey *k = RedisModule_OpenKey(sctx->redisCtx, doc->docKey, REDISMODULE_READ);
+  int rv = REDISMODULE_ERR;
+  if (!k || RedisModule_KeyType(k) != REDISMODULE_KEYTYPE_HASH) {
+    goto done;
+  }
+
+  size_t nitems = RedisModule_ValueLength(k);
+  if (nitems == 0) {
+    goto done;
+  }
+
+  Document_MakeStringsOwner(doc);
```

Comment:
We should come with a better name to this function, maybe take ownership?

Reply:
Well, it doesn't actually *take* ownership of the strings; it makes it own its own copy of the strings